components/object_store/src/obkv/mod.rs

```diff
@@ -0,0 +1,774 @@
+            MetaManager::try_new(client.clone()).map_err(|source| StoreError::Generic {
+                store: OBKV,
+                source: Box::new(source),
+            })?;
+        let upload_id = AtomicI64::new(0);
+        Ok(Self {
+            shard_manager,
+            meta_manager: Arc::new(meta_manager),
+            client,
+            upload_id,
+            part_size,
+        })
+    }
+
+    #[inline]
+    fn format_path(location: Option<&Path>) -> Path {
+        if let Some(path) = location {
+            if !path.as_ref().ends_with(DELIMITER) {
+                return Path::from(format!("{}{}", path.as_ref(), DELIMITER));
+            }
+            path.clone()
+        } else {
+            Path::from("")
+        }
+    }
+
+    #[inline]
+    pub fn pick_shard_table(&self, path: &Path) -> &str {
+        let mut hasher = DefaultHasher::new();
+        path.as_ref().as_bytes().hash(&mut hasher);
+        let hash = hasher.finish();
+        let index = hash % (self.shard_manager.shard_num as u64);
+        &self.shard_manager.table_names[index as usize]
+    }
+}
+
+impl<T: TableKv> ObkvObjectStore<T> {
+    async fn read_meta(&self, location: &Path) -> std::result::Result<ObkvObjectMeta, Error> {
+        let meta = self
+            .meta_manager
+            .read_meta(location)
+            .await
+            .box_err()
+            .context(ReadMeta {
+                path: location.as_ref().to_string(),
+            })?;
+
+        if let Some(m) = meta {
+            Ok(m)
+        } else {
+            MetaNotExists {
+                path: location.as_ref().to_string(),
+            }
+            .fail()
+        }
+    }
+
+    async fn get_internal(&self, location: &Path) -> std::result::Result<GetResult, Error> {
+        let meta = self.read_meta(location).await?;
+        let table_name = self.pick_shard_table(location);
+        print!("get_internal table_name:{table_name}");
+        let mut futures = FuturesOrdered::new();
+        for path in meta.parts {
+            let client = self.client.clone();
+            let table_name = table_name.to_string();
+            let future = async move {
+                match client.get(&table_name, path.as_bytes()) {
+                    Ok(res) => Ok(Bytes::from(res.unwrap())),
+                    Err(err) => Err(StoreError::Generic {
+                        store: OBKV,
+                        source: Box::new(err),
+                    }),
+                }
+            };
+            futures.push_back(future);
+        }
+
+        let boxed = futures.boxed();
+
+        Ok(GetResult::Stream(boxed))
+    }
+}
+
+#[async_trait]
+impl<T: TableKv> ObjectStore for ObkvObjectStore<T> {
+    async fn put(&self, location: &Path, bytes: Bytes) -> Result<()> {
+        let table_name = self.pick_shard_table(location);
+        let mut batch = T::WriteBatch::default();
+        batch.insert(location.as_ref().as_bytes(), bytes.as_ref());
```

Comment:
The bytes maybe very large, and in such case, shall we split the bytes into multiple parts for persisting in OBKV?

Reply:
In this implementation, I guess we can just reject the request whose the bytes size is too large.